### ISSUE TITLE
How to get a limited number of characters

### ISSUE THREAD

**Issue by thomasballinger** [thomasballinger/simplerepl#1]
@lovemac15 I've just added an example for this, see the first bullet point at https://github.com/thomasballinger/simplerepl. I'd love to hear what you think.

**Comment by ghost** [comment]
@thomasballinger that looks great! Thank you very much for taking the time to help me. I did, however, start learning the language last month so I'm still quite confused with it.

This is what I got so far: https://gist.github.com/lovemac15/c5e71e0b8aa428693e5b

My goal is to limit the number of characters a user can input to 'matricula'. I tried implementing your code in a couple of different ways:

This way first:

```
matricula = raw_input (term.move(4, 7) + "Matrícula: ")

with term.cbreak():
    while len(matricula) < 8:
        inp = term.inkey()
        if inp == '\n':
            sys.stdout.write('\n')
            sys.stdout.write('\r')
            sys.stdout.write('need more characters first!')
            sys.stdout.write(term.cuu(1))
        elif inp.code == term.KEY_DELETE:
            matricula = matricula[:-1]
        elif inp.is_sequence:
            pass
        else:
            matricula += inp
        sys.stdout.write('\r')
        sys.stdout.write(term.clear_eol())
        sys.stdout.write('.'*len(matricula))
        sys.stdout.flush()
        sys.stdout.write('\n')
        sys.stdout.flush()
        return matricula
```

That only showed the raw input then printed '.....'

The second way was by defining the function like in your example, then changing the raw input to:

```
matricula = get8chars
```

Which resulted in the following error: sqlite3.InterfaceError: Error binding parameter 0 - probably unsupported type.

As I said before, I'm a Python newbie so I'm pretty lost here. Any help would be appreciated! Thanks! :D

**Comment by thomasballinger** [comment]
Hi @lovemac15. In the first case, it was printing `...` because of the line that says `sys.stdout.write('.'*len(matricula))` - if you change it to `sys.stdout.write(matricula)` it will display what the user is typing. The second way you tried was a problem because you didn't call the function `get8chars` by using parentheses after it -- it should have been `matricula = get8chars()`.

I'm also confused about why you have `raw_input` in your example - I was thinking this function would replace the raw_input() call.

You might start with something simpler that will be easier to debug, like this:

```
import blessed

term = blessed.Terminal()

def get8chars():
    line = ''
    with term.cbreak():
        while len(line) < 8:
            line += term.inkey()
            print line
    return line

print get8chars()
```

it doesn't display very nicely, but you could get it working and then fix that with the cursor movement functionality it looks like you've been using.

**Comment by thomasballinger** [comment]
Also, from what I can tell from your example you need user input, but want to prevent the user from typing more than 8 characters? You might do something even simpler instead:

```
answer = raw_input('please enter 8 characters')
matricula = answer[:8]
```

or 

```
answer = ''
while len(answer) != 8:
    answer = raw_input('please enter 8 characters')
```

**Comment by lovemac15** [comment]
@thomasballinger yes, I'm not able to make it work by fixing the parenthesis mistake. My issue now is how to use term.move to position the input in the right position... How should I do that?

Also, I'm using this particular application for testing purposes only, my real goal is to use the character limitation on a different application that will ask for more than one input in the same line, somewhat like this:

    Nome: (user input)                                             Date of Birth: (user input)

So the first input should end before the second one, otherwise, the second one will be overlapped. So here's another issue. I should be able to type in less than 8 characters, but not more...

**Comment by ghost** [comment]
@thomasballinger yes, I was able to make it work by fixing the parenthesis mistake. My issue now is how to use term.move to position the input correctly... How should I do that?

Also, I'm using this particular application for testing purposes only, my real goal is to use the character limitation on a different application that will ask for more than one input in the same line, somewhat like this:

```
Nome: (user input)                                             Date of Birth: (user input)
```

The first input should end before the second one, otherwise, the second one will be overlapped.

I tried using:

```
answer = raw_input('please enter 8 characters')
matricula = answer[:8]
```

But that does not stop the input while it's being entered, which is what I need. Also, I want to be able to type in less than 8 characters, but not more...

**Comment by thomasballinger** [comment]
Great, now you'll want to move the cursor back up one line before the current line is printed. I suggest changing the print to:

sys.stdout.write(line)
sys.stdout.write('\r')
sys.stdout.flush()

This will write the line and then move the cursor back to the beginning of that line.

**Comment by ghost** [comment]
That only kept the cursor at the beginning of the line at all times.

**Comment by thomasballinger** [comment]
Ah, that's because '\r' moves the cursor to the beginning of the line. Try switching them:

```
sys.stdout.write('\r')
sys.stdout.write(line)
sys.stdout.flush()
```

**Comment by ghost** [comment]
That worked!

The real issue now is that I'm too much of a Python newbie to figure out how to change that function to make it do the following:

> Allow for 8 OR less characters to be entered
> Simply stop the input when the limit is reached, waiting for the user to either press enter and move forward, or edit the input
> Change the input position using blessed(or anything else for that matter...)

I don't want to waste any more of your time, you have taken a lot of it to help me with this issue and since you are a professional programmer, how much would you charge for editing that function for me? Assuming you are willing to...

Thanks :D

**Comment by thomasballinger** [comment]
Unfortunately I won't be able to do this, but I'm happy to help if you have questions about Blessed.

**Comment by ghost** [comment]
Oh ok. Thank you though! 

And I do, two actually:

> Is there any way to also use the key TAB to go to the next input?
> Is there any way to return to the previous input and edit it? For example, by pressing ALT+TAB...?

**Comment by thomasballinger** [comment]
I'd tackle those one at a time. It sounds like you want to, on ever
keystroke, check to see if the user typed tab instead of a normal
character, and if they did move the cursor to another spot on the screen
and when they type keys add those keys to a different word.

**Comment by ghost** [comment]
That's it!
